File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/StreamingReadFootgun.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*

@SuppressWarnings(Array("unused"))
object StreamingReadFootgun:
  var tx: Transactor = null // placeholder

  // start
  // WRONG: the cursor escapes the transaction — connection is already closed!
  def broken(): Cursor[String] =
    Fragment
      .of("SELECT name FROM users")
      .streamingQuery(PgTypes.text, 512)
      .transact(tx) // connection closes here, cursor is dead

  // CORRECT: process the cursor inside map, before the connection closes
  def correct(): Long =
    Fragment
      .of("SELECT name FROM users")
      .streamingQuery(PgTypes.text, 512)
      .map { cursor =>
        var count = 0L
        cursor.foreach(_ => count += 1)
        count
      }
      .transact(tx)
  // stop

File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/StreamingReadFootgun.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.*;

@SuppressWarnings("unused")
public class StreamingReadFootgun {
  Transactor tx = null; // placeholder

  // start
  // WRONG: the cursor escapes the transaction — connection is already closed!
  Cursor<String> broken() {
    return Fragment.of("SELECT name FROM users")
        .streamingQuery(PgTypes.text, 512)
        .transact(tx); // connection closes here, cursor is dead
  }

  // CORRECT: process the cursor inside map, before the connection closes
  long correct() {
    return Fragment.of("SELECT name FROM users")
        .streamingQuery(PgTypes.text, 512)
        .map(
            cursor -> {
              long count = 0;
              for (var name : cursor) count++;
              return count;
            })
        .transact(tx);
  }
  // stop
}
